webpmux.cpp

```diff
@@ -0,0 +1,90 @@
+#include "webpmux.hpp"
+#include <webp/encode.h>
+
+extern "C" {
+
+WebPAnimDecoder* webpmux_create_decoder(void* buf, size_t buf_len, void* info)
+{
+    auto i = static_cast<WebPAnimInfo*>(info);
+    WebPData data = {static_cast<const uint8_t*>(buf), buf_len};
+    WebPAnimDecoderOptions dec_options;
+    WebPAnimDecoderOptionsInit(&dec_options);
+    dec_options.color_mode = MODE_BGRA;
+    WebPAnimDecoder* dec = WebPAnimDecoderNew(&data, &dec_options);
+    if (dec == NULL)
+        return NULL;
+    WebPAnimDecoderGetInfo(dec, i);
+    return dec;
+}
+
+WebPAnimEncoder* webpmux_create_encoder(int width, int height)
+{
+    WebPAnimEncoderOptions enc_options;
+    WebPAnimEncoderOptionsInit(&enc_options);
+    return WebPAnimEncoderNew(width, height, &enc_options);
+}
+
+int webpmux_decoder_read_data(WebPAnimDecoder* dec, opencv_mat mat)
+{
+    auto m = static_cast<cv::Mat*>(mat);
+    int timestamp = 0;
+    uint8_t* buf = NULL;
+    int res = WebPAnimDecoderGetNext(dec, &buf, &timestamp);
+    if (!res)
+        return -1;
+    cv::Mat img(cv::Size(m->cols, m->rows), CV_8UC4, buf);
+    img.copyTo(*m);
+    return timestamp;
+}
+
+int webpmux_decoder_skip_frame(WebPAnimDecoder* dec)
+{
+    int timestamp = 0;
+    uint8_t* buf = NULL;
+    return WebPAnimDecoderGetNext(dec, &buf, &timestamp);
+}
+
+int webpmux_encoder_add_frame(WebPAnimEncoder* enc, opencv_mat mat, int timestamp, int quality)
+{
+    // TODO: Determine whether it's necessary to extract lossless flag
+    // Unfortunately, it appears to be non-trivial to determine whether a single
+    // frame in an animation was encoded losslessly, so it may be better to let
+    // the encoder decide for the sake of time, at the possible cost of intended
+    // quality. (We're already doing transformations anyway, so there's a loss
+    // of quality inherent to this process.)
+
+    auto m = static_cast<cv::Mat*>(mat);
+    WebPConfig config;
+    WebPConfigInit(&config);
+    config.lossless = false;
+    config.quality = quality;
+    WebPPicture frame;
+    WebPPictureInit(&frame);
+    frame.width = m->cols;
+    frame.height = m->rows;
+    frame.use_argb = true;
+    frame.argb = (uint32_t*)m->data;
+    frame.argb_stride = m->cols;
+    if (!WebPAnimEncoderAdd(enc, &frame, timestamp, &config)) {
+        // printf("failed to encode: %s\n", WebPAnimEncoderGetError(enc));
+        return frame.error_code;
+    }
+    return 0;
+}
+
+size_t webpmux_encoder_write(WebPAnimEncoder* enc, void* buf, size_t size, int timestamp)
+{
+    WebPAnimEncoderAdd(enc, NULL, timestamp, NULL);
+    WebPData webp_data;
+    WebPDataInit(&webp_data);
```

Comment:
Unfortunately it does not support custom buffers, but it doesn't need to resize the buffer, so I think this is just because it prefers to calculate the required size internally. It would be simple to patch it to respect the buffer passed in (ignoring the `const` qualifier :/), and have it complain about too little memory if the buffer's smaller than expected; saving the actual size back into the `WebPData` struct.

Also, if maximum memory allocation is your concern, libwebpmux does allow compiling with a `MALLOC_LIMIT` constant that will stop allocation past the specified limit.